File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/LoginActivity.java
package vinhlong.ditagis.com.khaosatdongho;

import android.content.Intent;
import android.os.Bundle;
import android.support.v7.app.AppCompatActivity;
import android.view.KeyEvent;
import android.view.View;
import android.widget.Button;
import android.widget.TextView;

import vinhlong.ditagis.com.khaosatdongho.async.LoginAsycn;
import vinhlong.ditagis.com.khaosatdongho.entities.DApplication;
import vinhlong.ditagis.com.khaosatdongho.entities.entitiesDB.User;
import vinhlong.ditagis.com.khaosatdongho.utities.CheckConnectInternet;
import vinhlong.ditagis.com.khaosatdongho.utities.Preference;


public class LoginActivity extends AppCompatActivity implements View.OnClickListener {
    private TextView mTxtUsername;
    private TextView mTxtPassword;
    private boolean isLastLogin;
    private TextView mTxtValidation;
    private DApplication dApplication;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_login);
        dApplication = (DApplication) getApplication();
        Button btnLogin = (findViewById(R.id.btnLogin));
        btnLogin.setOnClickListener(this);
        findViewById(R.id.txt_login_changeAccount).setOnClickListener(this);

        mTxtUsername = findViewById(R.id.txtUsername);
        mTxtPassword = findViewById(R.id.txtPassword);
        mTxtValidation = findViewById(R.id.txt_login_validation);
        create();
        login();
    }

    private void create() {

        String preference_userName = Preference.getInstance().loadPreference(getString(R.string.preference_username));

        //nếu chưa từng đăng nhập thành công trước đó
        //nhập username và password bình thường
        if (preference_userName == null || preference_userName.isEmpty()) {
            findViewById(R.id.layout_login_tool).setVisibility(View.GONE);
            findViewById(R.id.layout_login_username).setVisibility(View.VISIBLE);
            isLastLogin = false;
        }
        //ngược lại
        //chỉ nhập pasword
        else {
            isLastLogin = true;
            findViewById(R.id.layout_login_tool).setVisibility(View.VISIBLE);
            findViewById(R.id.layout_login_username).setVisibility(View.GONE);
        }

    }

    private void login() {
        if (!CheckConnectInternet.isOnline(this)) {
            mTxtValidation.setText(R.string.validate_no_connect);
            mTxtValidation.setVisibility(View.VISIBLE);
            return;
        }
        mTxtValidation.setVisibility(View.GONE);

        String userName, passWord;
        if (isLastLogin) {
            userName = Preference.getInstance().loadPreference(getString(R.string.preference_username));
            passWord = Preference.getInstance().loadPreference(getString(R.string.preference_password));
        } else {
            userName = mTxtUsername.getText().toString().trim();
            passWord = mTxtPassword.getText().toString().trim();
        }
        if (userName.length() == 0 || passWord.length() == 0) {
            handleInfoLoginEmpty();
            return;
        }
        LoginAsycn loginAsycn = new LoginAsycn(LoginActivity.this, output -> {
            if (output instanceof User) {
                User user = (User) output;
                dApplication.setUser(user);
                handleLoginSuccess(user);
            } else if (output instanceof String) {
                String s = (String) output;
                handleLoginFail(s);
            }
        });
        loginAsycn.execute(userName, passWord);
    }

    private void handleInfoLoginEmpty() {
        mTxtValidation.setText(R.string.info_login_empty);
        mTxtValidation.setVisibility(View.VISIBLE);
    }

    private void handleLoginFail(String message) {
        mTxtValidation.setText(message);
        mTxtValidation.setVisibility(View.VISIBLE);
    }

    private void handleLoginSuccess(User user) {
        Preference.getInstance().savePreferences(getString(R.string.preference_username),user.getUserName());
        Preference.getInstance().savePreferences(getString(R.string.preference_password), user.getPassWord());
        Preference.getInstance().savePreferences(getString(R.string.preference_displayname), user.getDisplayName());
        mTxtUsername.setText("");
        mTxtPassword.setText("");
        Intent intent = new Intent();
        setResult(RESULT_OK, intent);
        finish();
    }

    private void changeAccount() {
        mTxtUsername.setText("");
        mTxtPassword.setText("");

        Preference.getInstance().savePreferences(getString(R.string.preference_username), "");
        create();
    }

    @Override
    protected void onPostResume() {
        super.onPostResume();
        create();
    }

    @Override
    public void onClick(View view) {
        switch (view.getId()) {
            case R.id.btnLogin:
                login();
                break;
            case R.id.txt_login_changeAccount:
                changeAccount();
                break;
        }

    }

    @Override
    public boolean onKeyUp(int keyCode, KeyEvent event) {
        switch (keyCode) {
            case KeyEvent.KEYCODE_ENTER:
                if (mTxtPassword.getText().toString().trim().length() > 0) {
                    login();
                    return true;
                }
            default:
                return super.onKeyUp(keyCode, event);
        }
    }
}

File: app/src/main/java/vinhlong/ditagis/com/khaosatdongho/LoginActivity.kt
package vinhlong.ditagis.com.khaosatdongho

import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.support.v7.app.AppCompatActivity
import android.view.KeyEvent
import android.view.View
import android.widget.Button
import android.widget.TextView
import kotlinx.android.synthetic.main.activity_login.*

import vinhlong.ditagis.com.khaosatdongho.async.LoginAsycn
import vinhlong.ditagis.com.khaosatdongho.entities.DApplication
import vinhlong.ditagis.com.khaosatdongho.entities.entitiesDB.User
import vinhlong.ditagis.com.khaosatdongho.utities.CheckConnectInternet
import vinhlong.ditagis.com.khaosatdongho.utities.Constant
import vinhlong.ditagis.com.khaosatdongho.utities.DPreference


class LoginActivity : AppCompatActivity(), View.OnClickListener {
    private var mTxtUsername: TextView? = null
    private var mTxtPassword: TextView? = null
    private var isLastLogin: Boolean = false
    private var mTxtValidation: TextView? = null
    private lateinit var mApplication: DApplication

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_login)
        mApplication = application as DApplication
        val btnLogin = findViewById<Button>(R.id.btnLogin)
        btnLogin.setOnClickListener(this)
        findViewById<View>(R.id.txt_login_changeAccount).setOnClickListener(this)

        mTxtUsername = findViewById(R.id.txtUsername)
        mTxtPassword = findViewById(R.id.txtPassword)
        mTxtValidation = findViewById(R.id.txt_login_validation)
        create()
        login()
    }

    private fun create() {

        val preference_userName = DPreference.instance.loadPreference(Constant.PreferenceKey.USERNAME)

        //nếu chưa từng đăng nhập thành công trước đó
        //nhập username và password bình thường
        if (preference_userName == null || preference_userName.isEmpty()) {
            findViewById<View>(R.id.layout_login_tool).visibility = View.GONE
            findViewById<View>(R.id.layout_login_username).visibility = View.VISIBLE
            isLastLogin = false
        } else {
            isLastLogin = true
            findViewById<View>(R.id.layout_login_tool).visibility = View.VISIBLE
            findViewById<View>(R.id.layout_login_username).visibility = View.GONE
        }//ngược lại
        //chỉ nhập pasword

    }

    private fun login() {
        if (!CheckConnectInternet.isOnline(this)) {
            mTxtValidation!!.setText(R.string.validate_no_connect)
            mTxtValidation!!.visibility = View.VISIBLE
            return
        }
        mTxtValidation!!.visibility = View.GONE

        val userName: String?
        val passWord: String?
        if (isLastLogin) {
            userName = DPreference.instance.loadPreference(Constant.PreferenceKey.USERNAME)
            passWord = DPreference.instance.loadPreference(Constant.PreferenceKey.PASSWORD)
        } else {
            userName = mTxtUsername!!.text.toString().trim { it <= ' ' }
            passWord = mTxtPassword!!.text.toString().trim { it <= ' ' }
        }
        if (userName!!.isEmpty() || passWord!!.isEmpty()) {
            handleInfoLoginEmpty()
            return
        }
        mApplication.progressDialog.show(this@LoginActivity, container_login, "Đang đăng nhập...")
        val loginAsycn = LoginAsycn(this@LoginActivity, object : LoginAsycn.AsyncResponse {
            override fun processFinish(output: Any) {
                if (output is User) {
                    mApplication!!.user = output
                    handleLoginSuccess(output)
                } else if (output is String) {
                    handleLoginFail(output)
                }
                mApplication.progressDialog.dismiss()
            }


        })
        loginAsycn.execute(userName, passWord)
    }

    private fun handleInfoLoginEmpty() {
        mTxtValidation!!.setText(R.string.info_login_empty)
        mTxtValidation!!.visibility = View.VISIBLE
    }

    private fun handleLoginFail(message: String) {
        mTxtValidation!!.text = message
        mTxtValidation!!.visibility = View.VISIBLE
    }

    private fun handleLoginSuccess(user: User) {
        DPreference.instance.savePreferences(Constant.PreferenceKey.USERNAME, user.userName!!)
        DPreference.instance.savePreferences(Constant.PreferenceKey.PASSWORD, user.passWord!!)
        mTxtUsername!!.text = ""
        mTxtPassword!!.text = ""
        val intent = Intent()
        setResult(Activity.RESULT_OK, intent)
        finish()
    }

    private fun changeAccount() {
        mTxtUsername!!.text = ""
        mTxtPassword!!.text = ""

        DPreference.instance.savePreferences(Constant.PreferenceKey.USERNAME, "")
        create()
    }

    override fun onPostResume() {
        super.onPostResume()
        create()
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.btnLogin -> login()
            R.id.txt_login_changeAccount -> changeAccount()
        }

    }

    override fun onKeyUp(keyCode: Int, event: KeyEvent): Boolean {
        when (keyCode) {
            KeyEvent.KEYCODE_ENTER -> {
                if (mTxtPassword!!.text.toString().trim { it <= ' ' }.length > 0) {
                    login()
                    return true
                }
                return super.onKeyUp(keyCode, event)
            }
            else -> return super.onKeyUp(keyCode, event)
        }
    }
}
